pkg/api/job_runs.go

```diff
@@ -81,3 +86,182 @@ func JobsRunsReportFromDB(dbc *db.DB, filterOpts *filter.FilterOptions, release
+
+func runJobRunAnalysis(jobRun *models.ProwJobRun, compareRelease string,
+	testResultsFunc testResultsFunc) (apitype.ProwJobRunFailureAnalysis, error) {
+
+	logger := log.WithFields(log.Fields{
+		"func":     "jobRunAnalysis",
+		"jobRunID": jobRun.ID,
+	})
+
+	logger.WithField("url", jobRun.URL).Info("loaded prow job run for analysis")
+	logger.Infof("this job run has %d failed tests", len(jobRun.Tests))
+	logger.WithField("variants", jobRun.ProwJob.Variants).Debug("job variants")
+
+	response := apitype.ProwJobRunFailureAnalysis{
+		ProwJobRunID: jobRun.ID,
+		ProwJobName:  jobRun.ProwJob.Name,
+		ProwJobURL:   jobRun.URL,
+		Timestamp:    jobRun.Timestamp,
+		Tests:        []apitype.ProwJobRunTestFailureAnalysis{},
+		OverallRisk: apitype.FailureRisk{
+			Level:   apitype.FailureRiskLevelNone,
+			Reasons: []string{},
+		},
+		OpenBugs: jobRun.ProwJob.Bugs,
+	}
+
+	switch {
+
+	// Return early if no tests failed in this run:
+	case len(jobRun.Tests) == 0:
+		response.OverallRisk.Level = apitype.FailureRiskLevelNone
+		response.OverallRisk.Reasons = append(response.OverallRisk.Reasons,
+			"No test failures found in this job run.")
+		return response, nil
+
+	// Return early if we see mass test failures:
+	case len(jobRun.Tests) > 20:
+		response.OverallRisk.Level = apitype.FailureRiskLevelHigh
+		response.OverallRisk.Reasons = append(response.OverallRisk.Reasons,
+			fmt.Sprintf("%d tests failed in this run: High", len(jobRun.Tests)))
+		return response, nil
+	}
+
+	// Iterate each failed test, query it's pass rates by NURPs, find a matching variant combo, and
+	// see how often we've passed in the last week.
+	for _, ft := range jobRun.Tests {
+		logger.WithFields(log.Fields{
+			"testID": ft.TestID,
+			"name":   ft.Test.Name,
+		}).Debug("failed test")
+
+		testResult, err := testResultsFunc(
+			ft.Test.Name, compareRelease, ft.Suite.Name, jobRun.ProwJob.Variants)
+		if err != nil {
+			return response, err
+		}
+		if testResult != nil {
+			response.Tests = append(response.Tests, apitype.ProwJobRunTestFailureAnalysis{
+				Name: testResult.Name,
+				Risk: apitype.FailureRisk{
+					Level: getSeverityLevelForPassRate(testResult.CurrentPassPercentage),
+					Reasons: []string{
+						fmt.Sprintf("This test has passed %.2f%% of %d runs on release %s %v in the last week.",
+							testResult.CurrentPassPercentage, testResult.CurrentRuns, compareRelease, testResult.Variants),
+					},
+				},
+				OpenBugs: ft.Test.Bugs,
+			})
+		} else {
+			response.Tests = append(response.Tests, apitype.ProwJobRunTestFailureAnalysis{
+				Name: ft.Test.Name,
+				Risk: apitype.FailureRisk{
+					Level: apitype.FailureRiskLevelUnknown,
+					Reasons: []string{
+						fmt.Sprintf("Unable to find matching test results for variants: %v",
+							jobRun.ProwJob.Variants),
+					},
+				},
+			})
+		}
+	}
+
+	// Set the overall risk level for this job run to the highest we encountered:
+	var maxTestRiskReason string
```

Comment:
If we could calculate this each time we call the scoring api (if we add that) then we wouldn't have to post process

Reply:
Good idea, done!